pallets/external-validator-slashes/src/lib.rs

```diff
@@ -481,24 +511,66 @@ impl<T: Config> OnEraStart for Pallet<T> {
             }
         });
 
-        Self::confirm_unconfirmed_slashes(era_index);
+        Self::add_era_slashes_to_queue(era_index);
     }
 }
 
 impl<T: Config> Pallet<T> {
-    /// Apply previously-unapplied slashes on the beginning of a new era, after a delay.
-    fn confirm_unconfirmed_slashes(active_era: EraIndex) {
-        Slashes::<T>::mutate(&active_era, |era_slashes| {
-            log!(
-                log::Level::Debug,
-                "found {} slashes scheduled to be confirmed in era {:?}",
-                era_slashes.len(),
-                active_era,
-            );
-            for slash in era_slashes {
-                slash.confirmed = true;
+    fn add_era_slashes_to_queue(active_era: EraIndex) {
+        let mut slashes: VecDeque<_> = Slashes::<T>::take(&active_era).into();
+
+        UnreportedSlashesQueue::<T>::mutate(|queue| queue.append(&mut slashes));
+    }
+
+    fn process_slashes_queue_page() {
+        let mut slashes_to_send: Vec<_> = vec![];
+        let era_index = T::EraIndexProvider::active_era().index;
+
+        // prepare up to QueuedSlashesProcessedPerBlock slashes to be sent
+        for _ in 0..(T::QueuedSlashesProcessedPerBlock::get() as usize) {
+            let Some(slash) = UnreportedSlashesQueue::<T>::mutate(VecDeque::pop_front) else {
```

Comment:
(optimization): mutating storage inside a for loop sounds bad, let's try to read storage once and use drain to remove n elements @nanocryk:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a1269f106b287540fbb7fe2634f35b35

Reply:
Issue with drain is that it panics if the range goes out of bound. What we can do is place the mutate around the for loop so that it we perform multiple pop_front in memory and write the result once the limit is reached or pop_front fails.

Reply:
Ok didn't know it panics, the alternative sounds good